sys/net/network_layer/ng_sixlowpan/ng_sixlowpan.c

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (C) 2015 Martine Lenders <mlenders@inf.fu-berlin.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ */
+
+#include "kernel_types.h"
+#include "net/ng_netbase.h"
+#include "thread.h"
+#include "utlist.h"
+
+#include "net/ng_sixlowpan.h"
+#include "net/ng_sixlowpan/netif.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static kernel_pid_t _sixlowpan_pid = KERNEL_PID_UNDEF;
+static char _sixlowpan_stack[NG_SIXLOWPAN_STACK_SIZE];
+
+void _handle_receive(ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *payload;
+    uint8_t *dispatch;
+    ng_netreg_entry_t *entry;
+
+    LL_SEARCH_SCALAR(pkt, payload, type, NG_NETTYPE_SIXLOWPAN);
+
+    if ((payload == NULL) || (payload->size < 1)) {
+        DEBUG("6lo: Received packet has no 6LoWPAN payload\n");
+        ng_pktbuf_release(pkt);
+    }
+
+    dispatch = payload->data;
+
+    if (dispatch[0] == NG_SIXLOWPAN_UNCOMPRESSED) {
+        ng_pktsnip_t *sixlowpan;
+        DEBUG("6lo: received uncompressed IPv6 packet\n");
+        payload = ng_pktbuf_start_write(payload);
+
+        if (payload == NULL) {
+            DEBUG("6lo: can not get write access on received packet\n");
+#if defined(DEVELHELP) && defined(ENABLE_DEBUG)
+            ng_pktbuf_stats();
+#endif
+            ng_pktbuf_release(pkt);
+            return;
+        }
+
+        /* packet is uncompressed: just mark and remove the dispatch */
+        sixlowpan = ng_pktbuf_add(payload, payload->data, sizeof(uint8_t),
+                                  NG_NETTYPE_SIXLOWPAN);
+        LL_DELETE(pkt, sixlowpan);
+        ng_pktbuf_release(sixlowpan);
+    }
+    else {
+        DEBUG("6lo: dispatch %02x ... was not recognized\n",
+              dispatch[0]);
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    payload->type = NG_NETTYPE_IPV6;
+
+    entry = ng_netreg_lookup(NG_NETTYPE_IPV6, NG_NETREG_DEMUX_CTX_ALL);
+
+    if (entry == NULL) {
+        DEBUG("ipv6: No receivers for this packet found\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    ng_pktbuf_hold(pkt, ng_netreg_num(NG_NETTYPE_IPV6, NG_NETREG_DEMUX_CTX_ALL) - 1);
```

Comment:
I don't see why

> This receiver may immediately release the pkt for some reason, which may result in removing it before you can pass it to the second receiver.

is a problem. In that case `pkt->users` is decreased to 1 and we are on the same page as if there were only one receiver (the remaining one).

> Wouldn't it therefore be safer to not decrease by 1 here and do a ng_pktbuf_release(pkt) in the end of this function?

How would it be safer to create memory leaks?